File: projects/rccl/tools/device_linker/extract_fatbin.py
#!/usr/bin/env python3
"""
Extract and parse .hip_fatbin section from a shared library.

Usage: extract_fatbin.py <librccl.so> [output_dir]

The .hip_fatbin section may contain multiple components:
- CLANG_OFFLOAD_BUNDLE: Standard offload bundle with embedded device ELFs
- CCOB: Compressed Code OBject (used by --offload-compress)

Each component is written to a separate file in output_dir.
"""

import sys
import os
import struct
import subprocess
import tempfile
from pathlib import Path

CLANG_BUNDLE_MAGIC = b"__CLANG_OFFLOAD_BUNDLE__"
CCOB_MAGIC = b"CCOB"


def extract_hip_fatbin(so_path, output_path):
    """Extract .hip_fatbin section using llvm-objcopy."""
    llvm_objcopy = "/opt/rocm/llvm/bin/llvm-objcopy"
    if not os.path.exists(llvm_objcopy):
        llvm_objcopy = "llvm-objcopy"  # Try PATH
    
    result = subprocess.run(
        [llvm_objcopy, f"--dump-section=.hip_fatbin={output_path}", so_path],
        capture_output=True, text=True
    )
    if result.returncode != 0:
        raise RuntimeError(f"Failed to extract .hip_fatbin: {result.stderr}")
    
    if not os.path.exists(output_path):
        raise RuntimeError("llvm-objcopy succeeded but output file not created")
    
    return os.path.getsize(output_path)


def parse_clang_bundle(data, offset):
    """
    Parse a CLANG_OFFLOAD_BUNDLE starting at offset.
    Returns (bundle_info, end_offset) where bundle_info is a dict with metadata.
    """
    if data[offset:offset+24] != CLANG_BUNDLE_MAGIC:
        return None, offset
    
    pos = offset + 24
    num_entries = struct.unpack_from('<Q', data, pos)[0]
    pos += 8
    
    entries = []
    max_end = pos  # Track the furthest byte used by any entry
    
    for i in range(num_entries):
        entry_offset = struct.unpack_from('<Q', data, pos)[0]
        pos += 8
        entry_size = struct.unpack_from('<Q', data, pos)[0]
        pos += 8
        name_len = struct.unpack_from('<Q', data, pos)[0]
        pos += 8
        name = data[pos:pos+name_len].decode('utf-8', errors='replace').rstrip('\x00')
        pos += name_len
        
        entries.append({
            'name': name,
            'offset': entry_offset,  # Relative to start of bundle
            'size': entry_size
        })
        
        entry_end = entry_offset + entry_size
        if entry_end > max_end:
            max_end = entry_end
    
    # The bundle size is from the magic to the end of the last entry
    # Entries use offsets relative to the start of the bundle (offset parameter)
    bundle_size = max_end
    
    return {
        'type': 'CLANG_OFFLOAD_BUNDLE',
        'offset': offset,
        'size': bundle_size,
        'num_entries': num_entries,
        'entries': entries,
        'header_end': pos  # Where entry descriptors end
    }, offset + bundle_size


def parse_ccob(data, offset):
    """
    Parse a CCOB (Compressed Code OBject) starting at offset.
    Returns (ccob_info, end_offset).
    
    CCOB format (based on observation):
    - Magic: "CCOB" (4 bytes)
    - Version: uint16 (2 bytes)
    - Flags: uint16 (2 bytes)
    - Followed by zstd compressed data
    """
    if data[offset:offset+4] != CCOB_MAGIC:
        return None, offset
    
    # Read header
    version = struct.unpack_from('<H', data, offset + 4)[0]
    flags = struct.unpack_from('<H', data, offset + 6)[0]
    
    # CCOB doesn't have an explicit size field in the header
    # We need to find the end by looking for the next magic or EOF
    # Look for next CCOB or CLANG_BUNDLE_MAGIC or EOF
    pos = offset + 8
    end_pos = len(data)
    
    while pos < len(data) - 4:
        if data[pos:pos+4] == CCOB_MAGIC:
            end_pos = pos
            break
        if data[pos:pos+24] == CLANG_BUNDLE_MAGIC:
            end_pos = pos
            break
        pos += 1
    
    ccob_size = end_pos - offset
    
    return {
        'type': 'CCOB',
        'offset': offset,
        'size': ccob_size,
        'version': version,
        'flags': flags
    }, end_pos


def parse_fatbin(data):
    """
    Parse the entire .hip_fatbin data and return list of components.
    """
    components = []
    pos = 0
    
    while pos < len(data):
        # Skip zero padding
        while pos < len(data) and data[pos] == 0:
            pos += 1
        
        if pos >= len(data):
            break
        
        # Try to parse as CLANG_OFFLOAD_BUNDLE
        if pos + 24 <= len(data) and data[pos:pos+24] == CLANG_BUNDLE_MAGIC:
            info, new_pos = parse_clang_bundle(data, pos)
            if info:
                components.append(info)
                pos = new_pos
                continue
        
        # Try to parse as CCOB
        if pos + 4 <= len(data) and data[pos:pos+4] == CCOB_MAGIC:
            info, new_pos = parse_ccob(data, pos)
            if info:
                components.append(info)
                pos = new_pos
                continue
        
        # Unknown format - report error
        preview = data[pos:pos+20].hex() if pos + 20 <= len(data) else data[pos:].hex()
        raise RuntimeError(f"Unknown format at offset 0x{pos:x}: {preview}...")
    
    return components


def write_components(data, components, output_dir):
    """Write each component to a separate file."""
    os.makedirs(output_dir, exist_ok=True)
    
    for i, comp in enumerate(components):
        comp_type = comp['type']
        offset = comp['offset']
        size = comp['size']
        
        if comp_type == 'CLANG_OFFLOAD_BUNDLE':
            filename = f"bundle_{i}.bin"
        elif comp_type == 'CCOB':
            filename = f"ccob_{i}.bin"
        else:
            filename = f"unknown_{i}.bin"
        
        filepath = os.path.join(output_dir, filename)
        with open(filepath, 'wb') as f:
            f.write(data[offset:offset+size])
        
        print(f"  Wrote {filepath} ({size} bytes)")
        comp['file'] = filepath


def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <librccl.so> [output_dir]")
        sys.exit(1)
    
    so_path = sys.argv[1]
    output_dir = sys.argv[2] if len(sys.argv) > 2 else "fatbin_components"
    
    if not os.path.exists(so_path):
        print(f"Error: {so_path} not found")
        sys.exit(1)
    
    print(f"Extracting .hip_fatbin from {so_path}...")
    
    # Extract .hip_fatbin section
    with tempfile.NamedTemporaryFile(suffix='.bin', delete=False) as tmp:
        fatbin_path = tmp.name
    
    try:
        fatbin_size = extract_hip_fatbin(so_path, fatbin_path)
        print(f"  Extracted {fatbin_size} bytes ({fatbin_size/(1024*1024):.2f} MB)")
        
        # Read fatbin data
        with open(fatbin_path, 'rb') as f:
            data = f.read()
        
        # Parse components
        print(f"\nParsing .hip_fatbin...")
        components = parse_fatbin(data)
        
        print(f"\nFound {len(components)} component(s):")
        for i, comp in enumerate(components):
            print(f"\n  [{i}] {comp['type']}")
            print(f"      Offset: 0x{comp['offset']:x}")
            print(f"      Size: {comp['size']} bytes ({comp['size']/(1024*1024):.2f} MB)")
            
            if comp['type'] == 'CLANG_OFFLOAD_BUNDLE':
                print(f"      Entries: {comp['num_entries']}")
                for entry in comp['entries']:
                    print(f"        - {entry['name']}: offset=0x{entry['offset']:x}, size={entry['size']}")
            elif comp['type'] == 'CCOB':
                print(f"      Version: {comp['version']}")
                print(f"      Flags: 0x{comp['flags']:x}")
        
        # Check for unconsumed data
        total_consumed = sum(c['size'] for c in components)
        # Account for padding between components
        last_end = max(c['offset'] + c['size'] for c in components) if components else 0
        
        if last_end < len(data):
            # Check if remaining is just zeros
            remaining = data[last_end:]
            non_zero = sum(1 for b in remaining if b != 0)
            if non_zero > 0:
                print(f"\nWarning: {len(remaining)} bytes after last component ({non_zero} non-zero)")
            else:
                print(f"\nNote: {len(remaining)} bytes of zero padding after last component")
        
        # Write components
        print(f"\nWriting components to {output_dir}/...")
        write_components(data, components, output_dir)
        
        print(f"\nDone!")
        
    except Exception as e:
        print(f"\nError: {e}")
        sys.exit(1)
    finally:
        if os.path.exists(fatbin_path):
            os.unlink(fatbin_path)


if __name__ == "__main__":
    main()

File: projects/rccl/tools/device_linker/extract_fatbin.cpp
/*
 * Extract and parse .hip_fatbin section from a shared library.
 *
 * Usage: extract_fatbin <librccl.so> [output_dir]
 *
 * The .hip_fatbin section may contain multiple components:
 * - CLANG_OFFLOAD_BUNDLE: Standard offload bundle with embedded device ELFs
 * - CCOB: Compressed Code OBject (used by --offload-compress)
 *
 * Each component is written to a separate file in output_dir.
 *
 * Build: g++ -O2 -std=c++17 -o extract_fatbin extract_fatbin.cpp
 */

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cstdint>
#include <string>
#include <vector>
#include <fstream>
#include <filesystem>
#include <elf.h>

namespace fs = std::filesystem;

static const char CLANG_BUNDLE_MAGIC[] = "__CLANG_OFFLOAD_BUNDLE__";
static const char CCOB_MAGIC[] = "CCOB";

struct BundleEntry {
    std::string name;
    uint64_t offset;
    uint64_t size;
};

struct Component {
    std::string type;
    uint64_t offset;
    uint64_t size;
    // For CLANG_OFFLOAD_BUNDLE
    std::vector<BundleEntry> entries;
    // For CCOB
    uint16_t version;
    uint16_t flags;
};

// Read file into memory
std::vector<uint8_t> read_file(const std::string& path) {
    std::ifstream file(path, std::ios::binary | std::ios::ate);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + path);
    }
    
    size_t size = file.tellg();
    file.seekg(0);
    
    std::vector<uint8_t> data(size);
    file.read(reinterpret_cast<char*>(data.data()), size);
    
    if (!file) {
        throw std::runtime_error("Failed to read file: " + path);
    }
    
    return data;
}

// Write data to file
void write_file(const std::string& path, const uint8_t* data, size_t size) {
    std::ofstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot create file: " + path);
    }
    file.write(reinterpret_cast<const char*>(data), size);
}

// Extract .hip_fatbin section from ELF
std::vector<uint8_t> extract_hip_fatbin(const std::vector<uint8_t>& elf_data) {
    if (elf_data.size() < sizeof(Elf64_Ehdr)) {
        throw std::runtime_error("File too small for ELF header");
    }
    
    const auto* ehdr = reinterpret_cast<const Elf64_Ehdr*>(elf_data.data());
    
    if (memcmp(ehdr->e_ident, ELFMAG, SELFMAG) != 0) {
        throw std::runtime_error("Not an ELF file");
    }
    
    if (ehdr->e_ident[EI_CLASS] != ELFCLASS64) {
        throw std::runtime_error("Not a 64-bit ELF");
    }
    
    // Find section header string table
    if (ehdr->e_shstrndx == SHN_UNDEF || ehdr->e_shoff == 0) {
        throw std::runtime_error("No section header string table");
    }
    
    const auto* shdrs = reinterpret_cast<const Elf64_Shdr*>(elf_data.data() + ehdr->e_shoff);
    const auto& shstrtab_hdr = shdrs[ehdr->e_shstrndx];
    const char* shstrtab = reinterpret_cast<const char*>(elf_data.data() + shstrtab_hdr.sh_offset);
    
    // Find .hip_fatbin section
    for (uint16_t i = 0; i < ehdr->e_shnum; i++) {
        const char* name = shstrtab + shdrs[i].sh_name;
        if (strcmp(name, ".hip_fatbin") == 0) {
            uint64_t offset = shdrs[i].sh_offset;
            uint64_t size = shdrs[i].sh_size;
            
            if (offset + size > elf_data.size()) {
                throw std::runtime_error(".hip_fatbin section extends past end of file");
            }
            
            return std::vector<uint8_t>(
                elf_data.data() + offset,
                elf_data.data() + offset + size
            );
        }
    }
    
    throw std::runtime_error("No .hip_fatbin section found");
}

// Parse CLANG_OFFLOAD_BUNDLE at given offset
bool parse_clang_bundle(const std::vector<uint8_t>& data, size_t offset, Component& comp) {
    if (offset + 24 > data.size()) return false;
    if (memcmp(data.data() + offset, CLANG_BUNDLE_MAGIC, 24) != 0) return false;
    
    comp.type = "CLANG_OFFLOAD_BUNDLE";
    comp.offset = offset;
    
    size_t pos = offset + 24;
    
    if (pos + 8 > data.size()) return false;
    uint64_t num_entries;
    memcpy(&num_entries, data.data() + pos, 8);
    pos += 8;
    
    uint64_t max_end = pos;
    
    for (uint64_t i = 0; i < num_entries; i++) {
        if (pos + 24 > data.size()) return false;
        
        BundleEntry entry;
        memcpy(&entry.offset, data.data() + pos, 8);
        pos += 8;
        memcpy(&entry.size, data.data() + pos, 8);
        pos += 8;
        
        uint64_t name_len;
        memcpy(&name_len, data.data() + pos, 8);
        pos += 8;
        
        if (pos + name_len > data.size()) return false;
        entry.name = std::string(reinterpret_cast<const char*>(data.data() + pos), name_len);
        // Remove null terminator if present
        while (!entry.name.empty() && entry.name.back() == '\0') {
            entry.name.pop_back();
        }
        pos += name_len;
        
        comp.entries.push_back(entry);
        
        uint64_t entry_end = entry.offset + entry.size;
        if (entry_end > max_end) max_end = entry_end;
    }
    
    comp.size = max_end;
    return true;
}

// Parse CCOB at given offset
bool parse_ccob(const std::vector<uint8_t>& data, size_t offset, Component& comp) {
    if (offset + 8 > data.size()) return false;
    if (memcmp(data.data() + offset, CCOB_MAGIC, 4) != 0) return false;
    
    comp.type = "CCOB";
    comp.offset = offset;
    
    memcpy(&comp.version, data.data() + offset + 4, 2);
    memcpy(&comp.flags, data.data() + offset + 6, 2);
    
    // Find end by looking for next magic or EOF
    size_t pos = offset + 8;
    size_t end_pos = data.size();
    
    while (pos < data.size() - 4) {
        if (memcmp(data.data() + pos, CCOB_MAGIC, 4) == 0) {
            end_pos = pos;
            break;
        }
        if (pos + 24 <= data.size() && 
            memcmp(data.data() + pos, CLANG_BUNDLE_MAGIC, 24) == 0) {
            end_pos = pos;
            break;
        }
        pos++;
    }
    
    comp.size = end_pos - offset;
    return true;
}

// Parse entire fatbin
std::vector<Component> parse_fatbin(const std::vector<uint8_t>& data) {
    std::vector<Component> components;
    size_t pos = 0;
    
    while (pos < data.size()) {
        // Skip zero padding
        while (pos < data.size() && data[pos] == 0) {
            pos++;
        }
        
        if (pos >= data.size()) break;
        
        Component comp;
        
        // Try CLANG_OFFLOAD_BUNDLE
        if (parse_clang_bundle(data, pos, comp)) {
            components.push_back(comp);
            pos = comp.offset + comp.size;
            continue;
        }
        
        // Try CCOB
        if (parse_ccob(data, pos, comp)) {
            components.push_back(comp);
            pos = comp.offset + comp.size;
            continue;
        }
        
        // Unknown format
        fprintf(stderr, "Error: Unknown format at offset 0x%zx: ", pos);
        for (size_t i = 0; i < 20 && pos + i < data.size(); i++) {
            fprintf(stderr, "%02x", data[pos + i]);
        }
        fprintf(stderr, "...\n");
        throw std::runtime_error("Unknown format in fatbin");
    }
    
    return components;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <librccl.so> [output_dir]\n", argv[0]);
        return 1;
    }
    
    std::string so_path = argv[1];
    std::string output_dir = argc > 2 ? argv[2] : "fatbin_components";
    
    try {
        printf("Extracting .hip_fatbin from %s...\n", so_path.c_str());
        
        // Read ELF file
        auto elf_data = read_file(so_path);
        printf("  Read %zu bytes (%.2f MB)\n", elf_data.size(), elf_data.size() / (1024.0 * 1024.0));
        
        // Extract .hip_fatbin section
        auto fatbin_data = extract_hip_fatbin(elf_data);
        printf("  Extracted .hip_fatbin: %zu bytes (%.2f MB)\n", 
               fatbin_data.size(), fatbin_data.size() / (1024.0 * 1024.0));
        
        // Parse components
        printf("\nParsing .hip_fatbin...\n");
        auto components = parse_fatbin(fatbin_data);
        
        printf("\nFound %zu component(s):\n", components.size());
        
        for (size_t i = 0; i < components.size(); i++) {
            const auto& comp = components[i];
            printf("\n  [%zu] %s\n", i, comp.type.c_str());
            printf("      Offset: 0x%lx\n", (unsigned long)comp.offset);
            printf("      Size: %lu bytes (%.2f MB)\n", 
                   (unsigned long)comp.size, comp.size / (1024.0 * 1024.0));
            
            if (comp.type == "CLANG_OFFLOAD_BUNDLE") {
                printf("      Entries: %zu\n", comp.entries.size());
                for (const auto& entry : comp.entries) {
                    printf("        - %s: offset=0x%lx, size=%lu\n",
                           entry.name.c_str(), (unsigned long)entry.offset, (unsigned long)entry.size);
                }
            } else if (comp.type == "CCOB") {
                printf("      Version: %u\n", comp.version);
                printf("      Flags: 0x%x\n", comp.flags);
            }
        }
        
        // Check for unconsumed data
        if (!components.empty()) {
            uint64_t last_end = 0;
            for (const auto& c : components) {
                uint64_t end = c.offset + c.size;
                if (end > last_end) last_end = end;
            }
            
            if (last_end < fatbin_data.size()) {
                size_t remaining = fatbin_data.size() - last_end;
                size_t non_zero = 0;
                for (size_t i = last_end; i < fatbin_data.size(); i++) {
                    if (fatbin_data[i] != 0) non_zero++;
                }
                
                if (non_zero > 0) {
                    printf("\nWarning: %zu bytes after last component (%zu non-zero)\n", remaining, non_zero);
                } else {
                    printf("\nNote: %zu bytes of zero padding after last component\n", remaining);
                }
            }
        }
        
        // Write components
        printf("\nWriting components to %s/...\n", output_dir.c_str());
        fs::create_directories(output_dir);
        
        for (size_t i = 0; i < components.size(); i++) {
            const auto& comp = components[i];
            std::string filename;
            
            if (comp.type == "CLANG_OFFLOAD_BUNDLE") {
                filename = "bundle_" + std::to_string(i) + ".bin";
            } else if (comp.type == "CCOB") {
                filename = "ccob_" + std::to_string(i) + ".bin";
            } else {
                filename = "unknown_" + std::to_string(i) + ".bin";
            }
            
            std::string filepath = output_dir + "/" + filename;
            write_file(filepath, fatbin_data.data() + comp.offset, comp.size);
            printf("  Wrote %s (%lu bytes)\n", filepath.c_str(), (unsigned long)comp.size);
        }
        
        printf("\nDone!\n");
        return 0;
        
    } catch (const std::exception& e) {
        fprintf(stderr, "\nError: %s\n", e.what());
        return 1;
    }
}
